### PR TITLE
HDDS-8220. [Ozone-Streaming] Trigger volume check on IOException in StreamDataChannelBase

### PR DESCRIPTION
**Pull request by guohao-rosicky**
## What changes were proposed in this pull request?

KeyValueStreamDataChannel add write data exception handling

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8220

### CODE REVIEW

**Comment by kerneltime** [comment]
@hemantk-12 can you please take a look?

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/impl/KeyValueStreamDataChannel.java [hunk 1]
```diff
@@ -274,4 +286,8 @@ private static ContainerCommandRequestProto readPutBlockRequest(ByteBuffer b)
     }
     return request;
   }
+
+  private void volumeOnFailure() {
```
Comment (hemantk-12) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/impl/KeyValueStreamDataChannel.java:290]: nit: I don't think it is ideal to create function for one liner.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/impl/KeyValueStreamDataChannel.java [hunk 2]
```diff
@@ -198,8 +205,13 @@ void assertOpen() throws IOException {
   @Override
   public void close() throws IOException {
     if (closed.compareAndSet(false, true)) {
-      putBlockRequest.set(closeBuffers(buffers, super::writeFileChannel));
-      super.close();
+      try {
+        putBlockRequest.set(closeBuffers(buffers, super::writeFileChannel));
+        super.close();
```
Comment (hemantk-12) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/impl/KeyValueStreamDataChannel.java:210]: Should `super.close()` be out of the try-catch?

**Review by hemantk-12 — APPROVED** [review]
Overall looks good to me.

Left some minor comments.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/impl/KeyValueStreamDataChannel.java [hunk 3]
```diff
@@ -274,4 +286,8 @@ private static ContainerCommandRequestProto readPutBlockRequest(ByteBuffer b)
     }
     return request;
   }
+
+  private void volumeOnFailure() {
+    onFailure(getContainerData().getVolume());
```
Comment (szetszwo) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/impl/KeyValueStreamDataChannel.java:291]: @guohao-rosicky , thanks for working on this!

Let's move this method up and rename it to `checkVolume`.
```java
+++ b/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/impl/StreamDataChannelBase.java
@@ -22,6 +22,7 @@
 import org.apache.hadoop.hdds.scm.container.common.helpers.StorageContainerException;
 import org.apache.hadoop.ozone.container.common.helpers.ContainerMetrics;
 import org.apache.hadoop.ozone.container.common.impl.ContainerData;
+import org.apache.hadoop.ozone.container.common.utils.StorageVolumeUtil;
 import org.apache.ratis.statemachine.StateMachine;
 
 import java.io.File;
@@ -64,6 +65,10 @@ private FileChannel getChannel() {
     return randomAccessFile.getChannel();
   }
 
+  void checkVolume() {
+    StorageVolumeUtil.onFailure(containerData.getVolume());
+  }
+
   @Override
   public final void force(boolean metadata) throws IOException {
     getChannel().force(metadata);
```

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Pull request merged**